src/backends/asio/device.rs

```diff
@@ -0,0 +1,737 @@
+use std::{
+    borrow::Cow,
+    sync::{Arc, Mutex},
+};
+
+use asio_sys as asio;
+use num_traits::PrimInt;
+
+use crate::{
+    audio_buffer::{AudioMut, AudioRef},
+    device::{
+        AudioDevice, AudioDuplexDevice, AudioInputDevice, AudioOutputDevice, Channel, DeviceType,
+    },
+    duplex::AudioDuplexCallback,
+    stream::{
+        AudioCallbackContext, AudioInput, AudioInputCallback, AudioOutput, AudioOutputCallback,
+        StreamConfig,
+    },
+    timestamp::{self, Timestamp},
+    SendEverywhereButOnWeb,
+};
+
+use super::{error::AsioError, stream::AsioStream};
+
+/// The ASIO device.
+#[derive(Clone)]
+pub struct AsioDevice {
+    driver: Arc<asio::Driver>,
+    device_type: DeviceType,
+    asio_streams: Arc<Mutex<asio::AsioStreams>>,
+}
+
+impl AsioDevice {
+    /// Create a new ASIO device.
+    pub fn new(driver: Arc<asio::Driver>) -> Result<Self, AsioError> {
+        let is_input = driver.channels()?.ins > 0;
+        let is_output = driver.channels()?.outs > 0;
+        let device_type = match (is_input, is_output) {
+            (true, true) => DeviceType::DUPLEX,
+            (true, false) => DeviceType::INPUT,
+            (false, true) => DeviceType::OUTPUT,
+            // todo
+            (false, false) => return Err(AsioError::BackendError(asio::AsioError::NoDrivers)),
+        };
+        let asio_streams = Arc::new(Mutex::new(asio::AsioStreams {
+            input: None,
+            output: None,
+        }));
+        Ok(AsioDevice {
+            driver,
+            device_type,
+            asio_streams,
+        })
+    }
+
+    pub fn device_type(&self) -> DeviceType {
+        self.device_type
+    }
+
+    /// Create an input stream with the given configuration.
+    fn create_input_stream(&self, stream_config: StreamConfig) -> Result<usize, AsioError> {
+        let num_channels = stream_config.channels as usize;
+        let buffer_size = match stream_config.buffer_size_range {
+            (Some(min), Some(max)) if min == max => Some(min as i32),
+
+            _ => None,
+        };
```

Comment:
I think since we can control the exact buffer size, we should go with `min or max.min(512)` instead

Reply:
Not sure I understand. I thought when `min == max` then the user has specified a specific buffer size.

Reply:
Yes, however currently you're not respecting the conditions the user has provided in any other case (ex. `(None, Some(128)` or `(Some(256), Some(1024)`).

The expression I gate would translate in Rust as `min.or(max).unwrap_or(512)`. But I since saw that the buffer size you pass to `asio-sys` is actually an `Option`, so you can remove the `.unwrap_or` part.